feature/experimental/p4rt/otg_tests/google_discovery_protocol_packetin_test/google_discovery_protocol_packetin_test.go

```diff
@@ -257,14 +279,19 @@ func TestMain(m *testing.M) {
 }
 
 // configInterfaceDUT configures the interface with the Addrs.
-func configInterfaceDUT(i *oc.Interface, a *attrs.Attributes, dut *ondatra.DUTDevice) *oc.Interface {
+func configInterfaceDUT(i *oc.Interface, a *attrs.Attributes, dut *ondatra.DUTDevice, hasVlan bool) *oc.Interface {
 	i.Description = ygot.String(a.Desc)
 	i.Type = oc.IETFInterfaces_InterfaceType_ethernetCsmacd
 	if deviations.InterfaceEnabled(dut) {
 		i.Enabled = ygot.Bool(true)
 	}
-
-	s := i.GetOrCreateSubinterface(0)
+	var s *oc.Interface_Subinterface
+	if hasVlan && deviations.P4RTGdpRequiresDot1QSubinterface(dut) {
+		s = i.GetOrCreateSubinterface(1)
+		s.GetOrCreateVlan().GetOrCreateMatch().GetOrCreateSingleTagged().VlanId = &vlanID
+	} else {
+		s = i.GetOrCreateSubinterface(0)
+	}
```

Comment:
If the requirement for the test is to accept tagged frames for a specific VLAN, a sub-interface needs to configured with that VLAN-id. A sub-interface that is not configured for vlan-id, can receive only untagged frames. Therefore the code logic seems incorrect. The NOS implementations that need expliciti VLAN-id configuration are not the ones that need a deviation because their operation is accurate. It doesnt make sense for an implementation that is not configured for dot1q tagging to accept tagged frames. That implementation seems incorrect and should instead use a deviation.

Some relevance to this argument is also available in https://github.com/openconfig/public/blob/master/release/models/vlan/openconfig-vlan.yang#L264-L275

```
 grouping vlan-logical-config {
    description
      "VLAN related configuration that is part of subinterface
      (logical interface) configuration.  These are generally
      L3 VLANs with an id that is local or L2 VLANs attached with a wire
      transport service, such as a local cross-connect or an MPLS
      pseudo-wire.

      For ingress packets this VLAN configuration is used to match packets
      with an interface; for egress packets, the 'single-tagged' and
      'double-tagged' containers may be used to specify the VLAN identifiers
      that are added to packets as they leave the system.";
```

Reply:
While this is true for a normal ethernet frame being processed by the device, it may not be completely true for this scenario. 

In this scenario, the P4 Info set on the device should define the data plane behavior. And as per the P4Info defined, any packet with the GDP Ether type should be forwarded by the ASIC to the P4RT controller.

I am not sure which implementation is correct here. Maybe the P4 spec would have some answers.

Reply:
Another thing to note here is that, the bytes being sent can't actually be considered an Ethernet frame (even though it very much looks like one). It is a GDP (Google Discovery Protocol) packet and should never be processed like a ethernet frame.

Reply:
I agree with @sachendras that we should have subinterfaces enabled on the ROUTER to accept the tagged frame.
For Ethertype = 0x6007. It doesnt seem to be a random payload. Some reference to this EtherType show it as:
 #define ETHERTYPE_DEC_SCA       0x6007 
 ref: https://www.wireshark.org/docs/wsar_html/etypes_8h_source.html
or https://datatracker.ietf.org/doc/html/rfc2896

Even if we assume it as random payload, a dot1q padded frame has an ethertype of 0x8100 and the original ethertype is pushed as second ethertype. Ref: https://en.wikipedia.org/wiki/EtherType

Reply:
Well I am taking a full U turn now after getting more in to the depths of the expectations and talking to several folks internally. Let me first start w/ packet-out scenario as this is explicitly called out in the HLD

"The controller will inject packets that are to be forwarded by the device to support GDP through specifying the egress_port to which they should be forwarded. Ports are uniquely identified through the mechanism described in P4RT-4, and via the gRPC channel associated with the specified switching chip.
Additionally, the GDP packets may or may not have vlan-id set in them even if the interface is untagged. Device should be able to punt these packets up, and forward them out."

Having quoted that for the packet-out behavior from HLD, expectations for packet-in are also the same i.e.
The controller will program a set of rules for incoming GDP packets w/ ether-type 0x6007 for certain ingress interfaces w/ actions to either punt them to the controller or have them follow the FIB. These frames may carry vlan-id or not.  The device is expected to accept the packets even if the interface is untagged and follow the instructions programmed by the controller. 
Therefore, this is not a regular tagged packet that in traditional practice a NOS can accept only if the interface is configured for VLANs. Hence the code logic in this pull seems accurate and satisfies our production requirement. Any different behavior is a deviation.

Now, how to implement this in the NOS to cover for the latter i.e the packet-in expectations?

In the case of an ethernet frame with Dot1Q header, right after the "source mac address" comes the 2Byte TPID (Tag protocol identifier) of 0x8100. An NOS should be looking for the presence of this TPID. If one exists, ethertype of 0x6007 will appear after a 2Byte offset (offset to cover for VLAN ID etc). If after the source Mac the 2Byte TPID is missing then that is where the ethertype of 0x6007 can be found.

Hope this makes sense?

Reply:
I think packet out scenario is fine. Its only the packet in behavior which deviates from standard practice imo. I will leave it to NOS developers for any comment.